torch/csrc/jit/codegen/onednn/interface.cpp

```diff
@@ -0,0 +1,145 @@
+#include <torch/csrc/jit/jit_log.h>
+#include <oneapi/dnnl/dnnl_graph.hpp>
+#include <torch/csrc/jit/codegen/onednn/defer_size_check.h>
+#include <torch/csrc/jit/codegen/onednn/graph_fuser.h>
+#include <torch/csrc/jit/codegen/onednn/guard_shape.h>
+#include <torch/csrc/jit/codegen/onednn/interface.h>
+#include <torch/csrc/jit/codegen/onednn/kernel.h>
+#include <torch/csrc/jit/codegen/onednn/layout_propagation.h>
+#include <torch/csrc/jit/codegen/onednn/prepare_binary.h>
+#include <torch/csrc/jit/passes/decompose_ops.h>
+#include <torch/csrc/jit/passes/pass_manager.h>
+#include <torch/csrc/jit/passes/remove_mutation.h>
+#include <torch/csrc/jit/passes/tensorexpr_fuser.h>
+#include <torch/csrc/jit/runtime/custom_operator.h>
+#include <torch/csrc/jit/runtime/graph_executor.h>
+#include <torch/csrc/jit/runtime/operator_options.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace onednn {
+
+void fuseGraph(std::shared_ptr<Graph>& g) {
+  // Follow the process of the tensorexpr_fuser in profiling mode:
+  // Remove prim::profile nodes and embed the profile info directly in the
+  // IR in value types to avoid breaking the fusion patterns.
+  // Will add shape guard after LLGA optimization passes and
+  // wipe the tensor type information from the IR, so that it's not
+  // accidentally used by any other pass.
+
+  // We rely on the shape specialization and shape guard to ensure the validity
+  // of the cached compilation in the kernel, thus only support profiling mode.
+  // TODO: add check on oneDNNFusionGroup to ensure allShapesAreKnown on nodes
+  // to fuse: torch/csrc/jit/passes/tensorexpr_fuser.cpp: allShapesAreKnown
+  if (getProfilingMode()) {
+    GRAPH_DUMP(
+        "Before RemoveProfileNodesAndSpecializeTypes. Beginning of LLGA "
+        "optimization pass",
+        g);
+    RemoveProfileNodesAndSpecializeTypes(g);
+    GRAPH_DUMP(
+        "After RemoveProfileNodesAndSpecializeTypes. Before mutation removal",
+        g);
+
+    RemoveTensorMutation(g);
```

Comment:
We should pass in a lambda here so that we don't unmutate nodes which are unsupported in OneDNN (and also cpu for when this is enabled more globally)

Reply:
Thank you! I'll add something like this -


```
-    RemoveTensorMutation(g);
+    // Don't unmutate nodes which are unsupported in OneDNN Graph
+    RemoveTensorMutation(g,
+        [&](Node* node) {
+          return createOperator(node).kind() != dnnl::graph::op::kind::Wildcard;
+        });
```

Reply:
I tried this, which resulted in perf boost for mobilenet_v3_large, mobilenet_v2, resnet50 & timm_efficientnet, but regressions for other Vision models. I'll look into the reason.

<details>

```
    // Don't remove tensor-mutation for nodes unsupported by oneDNN Graph.
    // However, since currently, we've disabled single-op partitions for
    // most ops, they'd still end up being replaced by their functional
    // variants, because at this point, we've no way to tell if they'd
    // end up in single-op partitions. Maybe we can add another pass to
    // undo unnecessary inplace -> functional equivalent changes after
    // we'd know which ops, although supported by oneDNN Graph,
    // wouldn't be handled by it.
    RemoveTensorMutation(g, [](Node* nodeToFunctionalize) {
        static std::unordered_set<Symbol> supportedOps = {
            aten::add_,
            aten::mul_,
            aten::tanh_,
            aten::elu_,
            aten::relu_,
            aten::relu6_,
            aten::gelu_,
            aten::sqrt_,
            aten::sigmoid_,
            aten::hardtanh_,
            aten::abs_,
            aten::square_,
        };
        return supportedOps.count(nodeToFunctionalize->kind()) != 0;
      });

```

![image](https://user-images.githubusercontent.com/90872293/153663041-6e9aa527-b082-4171-8617-d3263a7ac605.png)

</details>

Reply:
> Were there any fusions that occurred in OneDNN that were happening before and now are not?

Since those ops are not mapped to oneDNN Graph, by not unmutating them, no fusion opportunities are lost from the oneDNN Graph perspective.

> It may be that globally unmutating - which NNC and NVFuser don't do yet - may be the best strategy for perf, however it'd be nice to keep changes local to this PR and make that decision a separate if we decide to go with it.

Thanks! Now we're only unmutating nodes that can be mapped to oneDNN Graph.